pkg/network/port_test.go

```diff
@@ -52,141 +50,136 @@ func TestMain(m *testing.M) {
-	defer ns.Close()
-
-	nsIno, err := kernel.GetInoForNs(ns)
-	require.NoError(t, err)
-
-	require.Eventually(t, func() bool {
-		initialPorts, err := ReadInitialState("/proc", TCP, true)
-		require.NoError(t, err)
-		for _, p := range ports[:2] {
-			if _, ok := initialPorts[PortMapping{testRootNs, p}]; !ok {
-				t.Errorf("PortMapping(testRootNs) returned false for port %d", p)
-				return false
-			}
+	kernel.WithNS(ns, func() error {
+		cmd := exec.Command("socat", "-d", "-d", "STDIO", address)
+		stderr, err := cmd.StderrPipe()
+		require.NoError(t, err, "error getting stderr pipe for command %s", cmd)
+		require.NoError(t, cmd.Start())
+		t.Cleanup(func() { cmd.Process.Kill() })
+		if port != 0 {
+			return nil
 		}
-		for _, p := range ports[2:] {
-			if _, ok := initialPorts[PortMapping{nsIno, p}]; !ok {
-				t.Errorf("PortMapping(test ns) returned false for port %d", p)
-				return false
+
+		scanner := bufio.NewScanner(stderr)
+		for scanner.Scan() {
+			matches := re.FindStringSubmatch(scanner.Text())
+			if len(matches) == 0 {
+				continue
 			}
-		}
 
-		if _, ok := initialPorts[PortMapping{testRootNs, 999}]; ok {
-			t.Errorf("expected PortMapping(testRootNs, 999) to not be in the map, but it was")
-			return false
-		}
-		if _, ok := initialPorts[PortMapping{nsIno, 999}]; ok {
-			t.Errorf("expected PortMapping(nsIno, 999) to not be in the map, but it was")
-			return false
+			require.Len(t, matches, 2)
+			_port, err := strconv.ParseUint(matches[1], 10, 16)
+			require.NoError(t, err)
+			port = uint16(_port)
+			break
 		}
 
-		return true
-	}, 3*time.Second, time.Second, "tcp/tcp6 ports are listening")
-}
-
-func TestReadInitialUDPState(t *testing.T) {
-	nsName := netlinktestutil.AddNS(t)
-	t.Cleanup(func() {
-		err := exec.Command("testdata/teardown_netns.sh").Run()
-		assert.NoError(t, err, "failed to teardown netns")
+		return nil
 	})
 
-	err := exec.Command("testdata/setup_netns.sh", nsName).Run()
-	require.NoError(t, err, "setup_netns.sh failed")
+	return port
+}
 
-	l := nettestutil.StartServerUDP(t, net.ParseIP("0.0.0.0"), 0)
-	t.Cleanup(func() {
-		l.Close()
+func TestReadInitialState(t *testing.T) {
+	t.Run("TCP", func(t *testing.T) {
+		testReadInitialState(t, "tcp")
 	})
-
-	l6 := nettestutil.StartServerUDP(t, net.ParseIP("::"), 0)
-	t.Cleanup(func() {
-		defer l6.Close()
+	t.Run("UDP", func(t *testing.T) {
+		testReadInitialState(t, "udp")
 	})
+}
 
-	conn, ok := l.(*net.UDPConn)
-	require.True(t, ok)
-	connl6, ok := l6.(*net.UDPConn)
-	assert.True(t, ok)
+func testReadInitialState(t *testing.T, proto string) {
+	var ns, rootNs netns.NsHandle
+	var err error
+	nsName := netlinktestutil.AddNS(t)
+	ns, err = netns.GetFromName(nsName)
+	require.NoError(t, err)
+	t.Cleanup(func() { ns.Close() })
+	rootNs, err = kernel.GetRootNetNamespace("/proc")
+	require.NoError(t, err)
+	t.Cleanup(func() { rootNs.Close() })
+
+	var protos []string
+	switch proto {
+	case "tcp":
+		protos = []string{"tcp4", "tcp6", "udp4", "udp6"}
```

Comment:
why do we still spin up udp servers when running in tcp mode and vice-versa? Do we do this on `main`?

Reply:
Yes, we spin up both udp and tcp to test if just listing tcp or udp doesn't include the other protocol ports.